access/email/plugindata_test.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2015-2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+var samplePluginData = PluginData{
+	RequestData: RequestData{
+		User:          "user-foo",
+		Roles:         []string{"role-foo", "role-bar"},
+		RequestReason: "foo reason",
+		ReviewsCount:  3,
+		Resolution:    Resolution{Tag: ResolvedApproved, Reason: "foo ok"},
+	},
+	EmailThreads: []EmailThread{
+		{Email: "E1", MessageID: "M1", Timestamp: "0000001"},
+		{Email: "E2", MessageID: "M2", Timestamp: "0000002"},
+	},
+}
+
+func TestEncodePluginData(t *testing.T) {
+	dataMap := EncodePluginData(samplePluginData)
+	assert.Len(t, dataMap, 7)
+	assert.Equal(t, "user-foo", dataMap["user"])
+	assert.Equal(t, "role-foo,role-bar", dataMap["roles"])
+	assert.Equal(t, "foo reason", dataMap["request_reason"])
+	assert.Equal(t, "3", dataMap["reviews_count"])
+	assert.Equal(t, "APPROVED", dataMap["resolution"])
+	assert.Equal(t, "foo ok", dataMap["resolve_reason"])
+	assert.Equal(t, "E1/0000001/M1,E2/0000002/M2", dataMap["email_threads"])
+}
+
+func TestDecodePluginData(t *testing.T) {
+	pluginData := DecodePluginData(map[string]string{
+		"user":           "user-foo",
+		"roles":          "role-foo,role-bar",
+		"request_reason": "foo reason",
+		"reviews_count":  "3",
+		"resolution":     "APPROVED",
+		"resolve_reason": "foo ok",
+		"email_threads":  "E1/0000001/M1,E2/0000002/M2",
+	})
+	assert.Equal(t, samplePluginData, pluginData)
+}
```

Comment:
I think it's worth combining these to reduce hard coded values and to make sure that the roundtrip works as expected. Also more of a nit, but you could just compare the `map[string]string`s directly.
```suggestion
func TestEncodeDecodePluginData(t *testing.T) {
	dataMap := EncodePluginData(samplePluginData)
	assert.Equal(t, map[string]string{
		"user":           "user-foo",
		"roles":          "role-foo,role-bar",
		"request_reason": "foo reason",
		"reviews_count":  "3",
		"resolution":     "APPROVED",
		"resolve_reason": "foo ok",
		"email_threads":  "E1/0000001/M1,E2/0000002/M2",
	}
	pluginData := DecodePluginData(dataMap)
	assert.Equal(t, samplePluginData, pluginData)
}
```'

Side note, is there a reason why you are using `assert` here? In `teleport`, we are transitioning to `testify/require` and it looks like teleport-plugins is also split between the two.

Reply:
1. Good idea. Combined tests.
2. Yes, it is between the to. I prefer require. `assert` is needed sometimes, in constructs like `next = next && assert.True(t, loaded)`. but there are not a lot of places like these. Let's just agree on require from now on.